### DISCUSSION TITLE
How to create and run executable of a monai code?

### DISCUSSION THREAD

**Discussion by Bhagii2508** [Project-MONAI/MONAI#4197] [Q&A] [answered]
Hello,
I have a segmentation code with monai architecture running perfectly with python execution, but when I try to create an executable and run it, it doesn't work. Could anyone please let me know how to fix this? Thanks

**Comment by rijobro** [comment]
What doesn't work? What error?

**Reply by Bhagii2508** [reply]
It is throwing me an import error on monai. I created the executable with the pyinstaller package. Also, I noticed on a lot of platforms that there is difficulty in creating executables with Pytorch workflow. Whereas the executables created with pyinstaller of Tensorflow architecture are working perfectly fine.

**Comment by plaresmedima** [comment]
Hi -I have the same problem. My program runs fine when executing from source but throws an error when I am running the executable created with pyinstaller:

ImportError: cannot import name 'CacheDataset' from partially initialized module 'monai.data' (most likely due to a circular import) (.\monai\data\__init__.pyc)

Does anyone know how to resolve it? My executable was running fine before adding the monai/pytorch feature.

**Reply by KumoLiu** [reply]
Hi @plaresmedima, maybe you can try some solutions introduced in https://stackoverflow.com/questions/69739699/pyinstaller-importerror-cannot-import-name-ccompiler-from-partially-initiali.
Hope it can help you, thanks!

**Comment by plaresmedima** [comment]
Hi thank you for the suggestion, but it seems the errors are specific to
monai. It tried adding the modules that have issues to hidden imports but
that does not help.

**Reply by wyli** [reply]
Thanks for raising the issue, could you help create a small example to demonstrate the issue? I can look into it. (Also a potential reason is that this line of load modules is not involved https://github.com/Project-MONAI/MONAI/blob/a89a0af10e70256fbe8641e71385ad2c2b3fb405/monai/__init__.py#L58)

**Comment by plaresmedima** [comment]
Dear Wenqi - just a quick note to say thank for your kind offer to look
into this. I have not had the time to write up a small example for you
unfortunately but I am keeping this on my todo list as it remains an
unsolved issue for us.

Prof. Steven Sourbron, PhD
Chair in Medical Imaging Physics
University of Sheffield, UK
*https
<https://www.sheffield.ac.uk/medicine/people/iicd/steven-sourbron>://www.sheffield.ac.uk/medicine/people/iicd/steven-sourbron
<https://www.sheffield.ac.uk/medicine/people/iicd/steven-sourbron>*
*Sheffield abdominal imaging research page*
<https://www.sheffield.ac.uk/medicine/research/research-themes/medical-imaging/medical-imaging-research/abdominal-imaging>
*Open source initiative in perfusion imaging* <http://www.osipi.org>
*Magnetic resonance imaging biomarkers for chronic kidney disease
<https://renalmri.org/>*

**Comment by VenturaFranklin** [comment, marked as answer]
I was able to get around this issue with the following code:
```python
import monai
datas = [(str(Path(monai.__file__).parent), "monai"),]
datas += PyInstaller.utils.hooks.collect_data_files("monai")

a = Analysis(
    ["{PATH_TO_YOUR_MAIN_FILE}"],
    pathex=["{PATH_TO_YOUR_MAIN_DIR}"],
    binaries=[],
    datas=datas,
    hiddenimports=hiddenimports,
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=None,
    noarchive=False,
)
```
Then you have your exe and collect configurations.

There might be better ways to do this, but this worked for me.
And of course you need to replace the path sections with the right paths for you.

> I have not had the time to write up a small example for you
unfortunately but I am keeping this on my todo list as it remains an
unsolved issue for us.

@plaresmedima Please, when asking for help, include your code even when it's not a small example. Copy and paste whatever you have and that is a million times better than nothing. And a small example is 1000 times better, but any code is better than nothing. I hope this helps, good luck!

**Reply by AmirV97** [reply]
Could u elaborate where i should add this code snippet? the 'a' part seems like part of the .spec file, but the first 2 lines seem like python code.

**Reply by VenturaFranklin** [reply]
@AmirV97 correct yes in the .spec file. You can have Python code in that file.

**Reply by AmirV97** [reply]
The code snippet uses the class Path in line 2, which is not imported. I'm using a .py script to get my exe file:
import PyInstaller.__main__
PyInstaller.__main__.run([
    'test__.spec',
])

test__.spec contain the code snippet. I'm getting the following error:
NameError: name 'Path' is not defined

I've added the line from pathlib import Path to both test__.spec and the .py script but i'm still getting the error.
